### ISSUE TITLE
Circularisation bearing

### ISSUE THREAD

**Issue by ElWanderer** [ElWanderer/kOS_scripts#140]
It's noticeable that if we have a rocket that takes a while to get the apoapsis up to the point where coasting is triggered, such that circularisation is a small burn, the final orbit is pretty accurate. However, the accuracy suffers the larger the circularisation burn.

The azimuth/bearing calculation assumes we burn for orbital velocity along the suggested bearing. However, what tends to happen is that we burn for roughly 3/4 of orbital velocity, then coast to apoapsis. When we plot a simple circularisation manuoevre at apoapsis, this "bakes in" whatever inclination we had got to. The longer we burn before coasting, the closer we get to the target inclination before coasting.

Solutions:
 * use a lower value for orbital velocity when calculating azimuth/bearing, in the hope we will hit the right inclination earlier during ascent, probably at some cost of delta-v
 * change the circularisation burn to correct the inclination, which sounds more complicated, but more realistic

Edit - the second of those is more complicated, though the code probably exists for it already, in the orbit matching library. Up until now, launch has not used the orbit functions; just a very simple calculation specific to circularisation, so it would be a big step. We can construct a target vector with the right velocity and heading (based on latitude of where the manoeuvre will be), and construct a manoeuvre to change the velocity to that.

**Comment by ElWanderer** [comment]
In lib_launch_common.ks, change:

```
FUNCTION launchBearing
{
  LOCAL lat IS SHIP:LATITUDE.
  LOCAL vo IS SHIP:VELOCITY:ORBIT.
  IF (LCH_I > 0 AND ABS(lat) < 90 AND MIN(LCH_I,180 - LCH_I) >= ABS(lat)) {
    LOCAL az IS ARCSIN( COS(LCH_I) / COS(lat) ).
    IF NOT LCH_AN { SET az TO mAngle(180 - az). }
    IF vo:MAG >= LCH_ORBIT_VEL { RETURN az. }
    LOCAL x IS (LCH_ORBIT_VEL * SIN(az)) - VDOT(vo,HEADING(90,0):VECTOR).
    LOCAL y IS (LCH_ORBIT_VEL * COS(az)) - VDOT(vo,HEADING(0,0):VECTOR).
    RETURN mAngle(90 - ARCTAN2(y, x)).
  } ELSE {
    IF LCH_I < 90 { RETURN 90. }
    ELSE { RETURN 270. }
  }
}
```
to a parameterised version:
```
FUNCTION launchBearing
{
  PARAMETER lat IS SHIP:LATITUDE, v0 IS SHIP:VELOCITY:ORBIT, v1_mag IS LCH_ORBIT_VEL.
  IF (LCH_I > 0 AND ABS(lat) < 90 AND MIN(LCH_I,180 - LCH_I) >= ABS(lat)) {
    LOCAL az IS ARCSIN( COS(LCH_I) / COS(lat) ).
    IF NOT LCH_AN { SET az TO mAngle(180 - az). }
    IF v0:MAG >= v1_mag { RETURN az. }
    LOCAL x IS (vt * SIN(az)) - VDOT(v0,HEADING(90,0):VECTOR).
    LOCAL y IS (vt * COS(az)) - VDOT(v0,HEADING(0,0):VECTOR).
    RETURN mAngle(90 - ARCTAN2(y, x)).
  } ELSE {
    IF LCH_I < 90 { RETURN 90. }
    ELSE { RETURN 270. }
  }
}
```
For circularisation we would pass in the latitude of the geoposition below the node, and a value of `v0` whose magnitude is greater than (or equal to) `v1_mag`, so that we just get back the azimuth.

Then we need to change the circularisation function (also in lib_launch_common.ks):

```
FUNCTION launchCirc
{
  IF NOT HASNODE {
    LOCAL m_time IS TIME:SECONDS + ETA:APOAPSIS.
    LOCAL v0 IS VELOCITYAT(SHIP,m_time):ORBIT:MAG.
    LOCAL v1 IS SQRT(BODY:MU/(BODY:RADIUS + APOAPSIS)).
    LOCAL n IS NODE(m_time, 0, 0, v1 - v0).
    addNode(n).
  }
  LOCAL ok IS execNode(TRUE).
  RETURN ok AND PERIAPSIS > BODY:ATM:HEIGHT.
}
```

to (requires lib_orbit_match.ks to get `nodeToVector()`):

```
FUNCTION launchCirc
{
  IF NOT HASNODE {
    LOCAL m_time IS TIME:SECONDS + ETA:APOAPSIS.
    LOCAL az IS launchBearing(BODY:GEOPOSITIONOF(POSITIONAT(SHIP,m_time)):LAT, V(0,0,0), -1).
    LOCAL v1 IS HEADING(az, 0):VECTOR:NORMALIZED * SQRT(BODY:MU/(BODY:RADIUS + APOAPSIS)).
    addNode(nodeToVector(v1, m_time)).
  }
  LOCAL ok IS execNode(TRUE).
  RETURN ok AND PERIAPSIS > BODY:ATM:HEIGHT.
}
```

**Comment by ElWanderer** [comment]
The code above isn't working properly. I suspect the result of `HEADING(azimuth, 0)` needs to be altered - it is probably returning a vector parallel to the horizon now, which is no longer parallel once we have got to apoapsis.

**Comment by ElWanderer** [comment]
This took a lot of effort and testing to get something that works. It didn't help that for a while I was generating a "vector towards the North Pole" that used `LATLNG(0,0)` (not the North Pole, but somewhere on the equator)... now it nails the inclination during circularisation, but the LAN tends to be a tenth or two of a degree off. That isn't surprising as it isn't controlled for, except by launch timing.

I've also slipped in a fix for a problem I suspected might exist (testing proved it did) - whether a launch is Northwards or Southwards was calculated and stored at launch, but it can actually change during a launch e.g. if you launch from Woomera (latitude 45 degrees) to an orbit with an inclination of 46 degrees, you are likely to launch and circularise in bearings that are either side of due East. Once a craft is ten times higher than the pitch-over altitude (2.5km for the default pitchover at 250m), the northwards/southwards bearing is re-evaluated based on the predicted track of the craft.
